finality-aleph/src/testing/data_store.rs

```diff
@@ -0,0 +1,645 @@
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN + 1].hash());
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN * 2 + 1].hash());
+
+    for _ in 1..=MAX_DATA_BRANCH_LEN {
+        let _message = test_handler
+            .assert_message_out("Did not receive message from Data Store")
+            .await;
+    }
+
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn hopeless_fork_at_the_boundary_goes_through() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10)
+        .await;
+    let fork_num = MAX_DATA_BRANCH_LEN + 2;
+    let forking_block = &blocks[fork_num];
+    let forks = test_handler
+        .build_branch_upon(&forking_block.hash(), MAX_DATA_BRANCH_LEN * 10)
+        .await;
+
+    test_handler.import_branch(blocks.clone(), false).await;
+
+    let honest_hopeless_fork = vec![
+        blocks[fork_num - 2].clone(),
+        blocks[fork_num - 1].clone(),
+        blocks[fork_num].clone(),
+        forks[0].clone(),
+    ];
+    let honest_hopeless_fork2 = vec![
+        blocks[fork_num - 2].clone(),
+        blocks[fork_num - 1].clone(),
+        blocks[fork_num].clone(),
+        forks[0].clone(),
+        forks[1].clone(),
+    ];
+    let malicious_hopeless_fork = vec![
+        blocks[fork_num - 2].clone(),
+        blocks[fork_num - 1].clone(),
+        blocks[fork_num].clone(),
+        forks[1].clone(),
+    ];
+    let malicious_hopeless_fork2 = vec![
+        blocks[fork_num - 2].clone(),
+        blocks[fork_num - 1].clone(),
+        blocks[fork_num].clone(),
+        blocks[fork_num + 1].clone(),
+        forks[1].clone(),
+    ];
+    let honest_data = vec![aleph_data_from_blocks(honest_hopeless_fork)];
+    let honest_data2 = vec![aleph_data_from_blocks(honest_hopeless_fork2)];
+    let malicious_data = vec![aleph_data_from_blocks(malicious_hopeless_fork)];
+    let malicious_data2 = vec![aleph_data_from_blocks(malicious_hopeless_fork2)];
+
+    test_handler.send_data(honest_data.clone());
+    test_handler.send_data(honest_data2.clone());
+    test_handler.send_data(malicious_data.clone());
+    test_handler.send_data(malicious_data2.clone());
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.finalize_block(&blocks[fork_num].hash());
+
+    let message = test_handler
+        .assert_message_out("Did not receive message from Data Store")
+        .await;
+    assert_eq!(message, malicious_data);
```

Comment:
probably I have a (hopefully) temporary blackout and I do not understand this testcase :| in particular, I have problems with the difference between malicious forks (they are not even valid proposals)

Reply:
Yes, this goes deep into the details of how malicious proposals should be dealt with. But in a nutshell, the definition of a hopeless fork proposal is:
``` 2) (Hopeless Fork) There exists a hash h_i on the branch, corresponding to height `num` in the chain, such that
    some block `b` of number `num` is finalized and `hash(b) != h`. This is simply a situation in which the proposal
   no matter whether honest or not, cannot be possibly be applied, as a conflicting block was already finalized.```

Malicious nodes might send arbitrary sequences of hashes as proposals, some of them we must accept as hopeless forks -- the reason is that sometimes we let (and must let) proposals through even if only a prefix of the branch has been imported (and we are not aware that the suffix are some random, malicious hashes).